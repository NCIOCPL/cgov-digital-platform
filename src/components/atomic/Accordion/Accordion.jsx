import React from 'react';
import PropTypes from 'prop-types';
import './Accordion.scss';

class Accordion extends React.Component {
  static propTypes = {
    bordered: PropTypes.bool,
    children: PropTypes.node,
    classes: PropTypes.string,
    startCollapsed: PropTypes.bool
  };

  static defaultProps = {
    bordered: false,
    classes: '',
    startCollapsed: false
  };

  constructor(props) {
    super(props);

    this.state = {
      activeIndex: props.startCollapsed ? -1 : 0,
    };
  }

  //  Before the component mounts, loop through the children to see if any of
  //  them have the `expanded` prop.  This allows consumers of this class to
  //  specify which specific item they want expanded when the component is
  //  rendered.
  //
  //  If no items have `expanded` passed as a prop, then the default behavior
  //  will take effect.
  //
  //  If more than one item has `expanded`, the last one wins and a console
  //  warning will be printed.
  _initialExpand() {
    let foundExpandedItem = false;
    let children = React.Children.toArray(this.props.children);
    for (let i = 0; i < children.length; i++) {
      let child = children[i];
      if (child.props.expanded) {
        if (foundExpandedItem) {
          console.warn(
            'Only one AccordionItem can be expanded at a time. You have marked more than one for expansion by default.'
          ); // eslint-disable-line no-console
        }
        this.setState({ activeIndex: i });
        foundExpandedItem = true;
      }
    }
  }
  //  Sets the active accordion item.
  //
  //  This method is assigned as the `action` of child {@link AccordionItem}
  //  components.  When the user clicks on a header of one of the accordion items,
  //  it will invoke this method with its index.  We can simply update our state,
  //  which will cause the children to be re-rendered so that the correct item
  //  has its content displayed.
  //
  //  To collapse all the items, pass -1 as the value.
  //
  //  @param {Number} index The item index that should be the active item.
  //
  componentDidMount() {
    this._initialExpand();
  }

  setActiveItem(index) {
    this.setState({ activeIndex: index });
  }

  render() {
    let index = 0;
    let children = React.Children.map(this.props.children, child => {
      let i = index++;
      return React.cloneElement(child, {
        accordionIndex: i,
        action: this.setActiveItem.bind(this),
        expanded: i === this.state.activeIndex,
      });
    });
    return (
      <div
        className={`cts-accordion ${
          this.props.bordered ? 'cts-accordion--bordered' : ''
        } ${this.props.classes}`}
      >
        {children}
      </div>
    );
  }
}

export default Accordion;
