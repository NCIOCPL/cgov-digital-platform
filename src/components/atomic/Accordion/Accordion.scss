$cts-accordion-border: #ddd;
$cts-accordion-heading-bg-odd: #f2f2f2;
$cts-accordion-heading-toggle-odd: #006F8A;


.cts-accordion {
  margin-bottom: 0;
  margin-top: 0;
  list-style-type: none;
  padding-left: 0;
  color: #1b1b1b;
  margin: 0;
  padding: 0;
  width: 100%;
  font-size: 1em;
  line-height: 1.52155;

  &--bordered {
    .cts-accordion__heading {
      border-top: 1px solid $cts-accordion-border;
      border-left: 1px solid $cts-accordion-border;
      border-right: 1px solid $cts-accordion-border;
      margin-bottom: 0;
      &[aria-expanded='false'] {
        border-bottom: 1px solid $cts-accordion-border;
      }
    }
    .cts-accordion__content {
      border-bottom: 1px solid $cts-accordion-border;
      border-left: 1px solid $cts-accordion-border;
      border-right: 1px solid $cts-accordion-border;
      padding-bottom: 1rem;

      + .cts-accordion__heading {
        border-top: 0;
      }
    }
  }

  > li {
    margin-bottom: 0;
    max-width: unset;
  }

  > ul li ul {
    list-style: disc;
  }

  > ul li ul > li > ul {
    list-style: circle;
  }

  > ul li ul > li > ul > li > ul {
    list-style: square;
  }

  + .cts-accordion,
  + .cts-accordion--bordered {
    margin-top: 0.5rem;
  }

  &__heading {
    font-size: 1rem;
    line-height: 1;
    margin: 0;
  }

  &__content {
    background: white;
    margin-top: 0;
    overflow: auto;
    padding: 1rem 1.25rem calc(1rem - 0.25rem) 1.25rem;
    transition: max-height 1s;
    height: auto;
    max-height: 10000px;
    
    &[aria-hidden='true'] {
      display: none;
      max-height: 0;
    }
    > *:first-child {
      margin-top: 0;
    }
    > *:last-child {
      margin-bottom: 0;
    }
    :not(:first-child) {
      margin-top: 0.5rem;
    }
  }

  h2{
    &:nth-of-type(odd){
      .cts-accordion__button {
        background: $cts-accordion-heading-bg-odd;

        &::after {
          background: white;
          color: $cts-accordion-heading-toggle-odd;
        }
        &[aria-expanded='false'] {
          &::after{
            background: $cts-accordion-heading-toggle-odd;
            color: white;
          }
        }

      }
    }
  }

  &__button {
    -moz-osx-font-smoothing: auto;
    -webkit-font-smoothing: subpixel-antialiased;
    background: #dbf4f6;
    color: #606060;
    text-decoration: underline;
    border: 0;
    border-radius: 0;
    box-shadow: none;
    font-weight: normal;
    font-size: 1.125em;
    margin: 0;
    padding: 0;
    text-align: left;
    cursor: pointer;
    display: inline-block;
    font-weight: bold;
    margin: 0;
    padding: 14.5px 70px 14.5px 20px;
    text-decoration: none;
    width: 100%;
    position: relative;

    &::after {
      content: '\2212';
      display: inline-block;
      width: 50px;
      height: 50px;
      background: white;
      position: absolute;
      right: 0;
      top: 0;
      text-align: center;
      color: #b2dcdf;
      font-size: 1.9em;
    }

    &[aria-expanded='false'] {
      &::after{
        content: '+';
        background: #b2dcdf;
        color: white;
      }
    }

    &:active {
      color: #162e51;
    }
    &:focus {
      outline: 0.25rem solid #2491ff;
      outline-offset: 0;
    }
    &:visited {
      color: #562b97;
    }

  }
}
