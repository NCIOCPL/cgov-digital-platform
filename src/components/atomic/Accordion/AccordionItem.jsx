import React from 'react';
import PropTypes from 'prop-types';
import Utilities from '../../../utilities/utilities';

class AccordionItem extends React.Component {
  static propTypes = {
    accordionIndex: PropTypes.number,
    action: PropTypes.func,
    children: PropTypes.node,
    expanded: PropTypes.bool,
    titleCollapsed: PropTypes.string,
    titleExpanded: PropTypes.string
  };

  static defaultProps = {
    action: function() {},
    expanded: false,
    title: '',
  };

  constructor(props) {
    super(props);

    this.state = {
      uuid: '',
    };
  }

  //  Generates a unique ID for this element using the {@link Utilities}
  //  function {@link uniqueIdForComponent}.
  //
  //  This must happen after the component has been mounted because the
  //  {@link uniqueIdForComponent} method will check to see if the DOM element
  //  has a `data-reactid` attribute and use that if present.  This helps keep
  //  continuity between what was rendered on the server and what the client
  //  sees.
  //
  //  After retrieving the appropriate unique ID, we set our state's `uuid`
  //  variable to that ID which will cause the component to re-render.

  componentDidMount() {
    let id = Utilities.uniqueIdForComponent(this);
    this.setState({ uuid: id });
  };

  //  Click handler for the title element.
  //
  //  This method calls the method defined in our `action` prop (usually set
  //  by the parent {@link Accordion} class) with our index to make ourselves
  //  the active element and show our content.
  //
  //  If this item is currently active, it will pass -1 as the value to the
  //  parent {@link Accordion} so that it will collapse the item.
  makeActive = () => {
    if (this.props.expanded) {
      this.props.action(-1);
    } else {
      this.props.action(this.props.accordionIndex);
    }
  };

  //  Renders the title element of this accordion item.
  //
  //  This element will act as a button and invoke our {@link makeActive}
  //  method whenever it is clicked.
  //
  //  See the class documentation for more information on how to supply the value
  //  for this element.
  //
  //  @returns {Node} The rendered DOM node.
  renderTitleElement() {
    let element;
    if (this.props.titleCollapsed.length > 0) {
      element = <span>{(this.props.expanded && this.props.titleExpanded)? this.props.titleExpanded : this.props.titleCollapsed}</span>;
    } else {
      if (React.Children.count(this.props.children) !== 2) {
        throw new Error('Either a title or 2 child elements must be supplied.');
      }
      let children = React.Children.toArray(this.props.children);
      element = children[0];
    }
    return (
      <h2 className="cts-accordion__heading" aria-expanded={this.props.expanded}>
        <button
          className="cts-accordion__button"
          aria-expanded={this.props.expanded}
          aria-controls={`${this.state.uuid}-content`}
          onClick={this.makeActive}
        >
          {element}
        </button>
      </h2>
    );
  };

  //  Renders our content element
  //
  //  @returns {Node} The rendered DOM node

  renderContentElement() {
    let children = React.Children.toArray(this.props.children);
    let element = children.length === 2 ? children[1] : children[0];
    return (
      <div
        id={`${this.state.uuid}-content`}
        className="cts-accordion__content"
        aria-hidden={!this.props.expanded}
      >
        {element}
      </div>
    );
  };

  render() {
    // Ensure there are only 2 children.
    if (React.Children.count(this.props.children) > 2) {
      throw new Error(
        'AccordionItem elements must have no more than 2 children.'
      );
    }

    return (
      <React.Fragment>
        {this.renderTitleElement()}
        {this.renderContentElement()}
      </React.Fragment>
    );
  };
}

export default AccordionItem;
