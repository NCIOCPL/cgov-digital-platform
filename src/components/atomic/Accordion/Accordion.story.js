import React from 'react';
import generateStories from '../../../../.storybook/generateStories';
import Accordion from './Accordion';
import AccordionItem from './AccordionItem';

generateStories({
  component: { Accordion },
  storyName: 'Elements | Accordion',
  defaultProps: {
    bordered: false
  },
  defaultChildren: (
    <React.Fragment>
      <AccordionItem title="First Amendment" expanded>
        <p>Congress shall make no law respecting an establishment of ...</p>
      </AccordionItem>
      <AccordionItem>
        <span>Second Amendment</span>
        <p>A well regulated Militia, being necessary to the security ...</p>
      </AccordionItem>
    </React.Fragment>
  ),
  stories: [
    {
      name: 'default',
    },
  ],
});
