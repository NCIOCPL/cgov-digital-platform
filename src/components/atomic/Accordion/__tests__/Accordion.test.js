import React from 'react';
import { mount } from 'enzyme';
import Accordion from '../Accordion.jsx';
import AccordionItem from '../AccordionItem.jsx';

jest.unmock('../Accordion.jsx');
jest.unmock('../AccordionItem.jsx');

describe('Accordion', () => {
  let accordion = null;

  beforeEach(function() {
    accordion = mount(
      <Accordion>
        <AccordionItem title="First Amendment">
          <p>Congress shall make no law respecting an establishment of ...</p>
        </AccordionItem>
        <AccordionItem>
          <span>Second Amendment</span>
          <p>A well regulated Militia, being necessary to the security ...</p>
        </AccordionItem>
      </Accordion>
    );
  });

  afterEach(() => {
    accordion.unmount();
  });

  // it('should render correctly', () => {
  //   expect(accordion).toMatchSnapshot();
  // });

  it('is a div element', () => {
    expect(accordion.type()).toEqual(Accordion);
    expect(accordion.getDOMNode().tagName.toLowerCase()).toEqual('div');
  });

  it('has two items', () => {
    expect(accordion.find('AccordionItem').length).toEqual(2);
  });

  it('supports setting the title as a prop', () => {
    let text = 'First Amendment';
    expect(
      accordion
        .find('AccordionItem')
        .at(0)
        .prop('title')
    ).toEqual(text);
    expect(
      accordion
        .find('AccordionItem')
        .at(0)
        .find('button span')
        .first()
        .text()
    ).toEqual(text);
  });

  it('supports setting the title as a child element', () => {
    let text = 'Second Amendment';
    expect(
      accordion
        .find('AccordionItem')
        .at(1)
        .find('button span')
        .first()
        .text()
    ).toEqual(text);
  });

  it('shows the content for the first item by default', () => {
    expect(accordion.state('activeIndex')).toEqual(0);
    expect(
      accordion
        .find('AccordionItem')
        .at(0)
        .prop('expanded')
    ).toBeTruthy();
  });

  it('should throw an exception if no title or 2 elements are provided', () => {
    expect(() => {
      mount(
        <Accordion>
          <AccordionItem>
            <p>Congress shall make no law respecting an establishment of ...</p>
          </AccordionItem>
        </Accordion>
      );
    }).toThrowError();
  });

  it('show throw if child has more than 2 child nodes', () => {
    expect(() => {
      mount(
        <Accordion>
          <AccordionItem>
            <span>Second Amendment</span>
            <p>A well regulated Militia, being necessary to the security ...</p>
            <p>Unsupported 3rd child</p>
          </AccordionItem>
        </Accordion>
      );
    }).toThrowError();
  });

  it("should have no expanded items if 'startCollapsed' is passed as a prop", () => {
    accordion = mount(
      <Accordion startCollapsed>
        <AccordionItem title="First Amendment">
          <p>Congress shall make no law respecting an establishment of ...</p>
        </AccordionItem>
        <AccordionItem>
          <span>Second Amendment</span>
          <p>A well regulated Militia, being necessary to the security ...</p>
        </AccordionItem>
      </Accordion>
    );
    expect(
      accordion
        .find('AccordionItem')
        .at(0)
        .prop('expanded')
    ).toBeFalsy();
    expect(
      accordion
        .find('AccordionItem')
        .at(1)
        .prop('expanded')
    ).toBeFalsy();
  });

  // it('should allow a specific item to start expanded', () => {
  //   accordion = mount(
  //     <Accordion>
  //       <AccordionItem title='First Amendment'>
  //         <p>Congress shall make no law respecting an establishment of ...</p>
  //       </AccordionItem>
  //       <AccordionItem expanded>
  //         <span>Second Amendment</span>
  //         <p>A well regulated Militia, being necessary to the security ...</p>
  //       </AccordionItem>
  //     </Accordion>
  //   );
  //   expect(accordion.find('AccordionItem').at(0).prop('expanded')).toBeFalsy();
  //   expect(accordion.find('AccordionItem').at(1).prop('expanded')).toBeTruthy();

  // });
});
