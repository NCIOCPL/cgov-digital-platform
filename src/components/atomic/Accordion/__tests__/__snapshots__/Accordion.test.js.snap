// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly without errors 1`] = `
<Accordion
  bordered={false}
  startCollapsed={false}
>
  <div
    className="cts-accordion "
  >
    <AccordionItem
      accordionIndex={0}
      action={[Function]}
      expanded={true}
      key=".0"
      title="First Amendment"
    >
      <h2
        aria-expanded={true}
        className="cts-accordion__heading"
      >
        <button
          aria-controls="component-unique-id-0-content"
          aria-expanded={true}
          className="cts-accordion__button"
          onClick={[Function]}
        >
          <span>
            First Amendment
          </span>
        </button>
      </h2>
      <div
        aria-hidden={false}
        className="cts-accordion__content"
        id="component-unique-id-0-content"
      >
        <p
          key=".0"
        >
          Congress shall make no law respecting an establishment of ...
        </p>
      </div>
    </AccordionItem>
    <AccordionItem
      accordionIndex={1}
      action={[Function]}
      expanded={false}
      key=".1"
      title=""
    >
      <h2
        aria-expanded={false}
        className="cts-accordion__heading"
      >
        <button
          aria-controls="component-unique-id-1-content"
          aria-expanded={false}
          className="cts-accordion__button"
          onClick={[Function]}
        >
          <span
            key=".0"
          >
            Second Amendment
          </span>
        </button>
      </h2>
      <div
        aria-hidden={true}
        className="cts-accordion__content"
        id="component-unique-id-1-content"
      >
        <p
          key=".1"
        >
          A well regulated Militia, being necessary to the security ...
        </p>
      </div>
    </AccordionItem>
  </div>
</Accordion>
`;
